crates/sui-indexer-restorer/src/snapshot.rs

```diff
@@ -0,0 +1,214 @@
+            db,
+            next_checkpoint_after_epoch,
+        })
+    }
+
+    pub async fn restore(&mut self) -> Result<(), Error> {
+        info!(
+            "Starting snapshot restore from epoch {}",
+            self.restore_args.start_epoch
+        );
+        let (sha3_digests, num_part_files) = self.snapshot_reader.compute_checksum().await?;
+        let (_abort_handle, abort_registration) = AbortHandle::new_pair();
+        let (input_files, epoch_dir, remote_object_store, _concurrency) =
+            self.snapshot_reader.export_metadata().await?;
+        let owned_input_files: Vec<(u32, (u32, FileMetadata))> = input_files
+            .into_iter()
+            .map(|(bucket, (part_num, metadata))| (*bucket, (part_num, metadata.clone())))
+            .collect();
+
+        self.restore_object_infos(
+            abort_registration,
+            owned_input_files,
+            epoch_dir,
+            remote_object_store,
+            sha3_digests,
+            num_part_files,
+            self.restore_args.concurrency,
+        )
+        .await?;
+        info!(
+            "Finished snapshot restore from epoch {}",
+            self.restore_args.start_epoch
+        );
+        Ok(())
+    }
+
+    async fn restore_object_infos(
+        &self,
+        abort_registration: AbortRegistration,
+        input_files: Vec<(u32, (u32, FileMetadata))>,
+        epoch_dir: Path,
+        remote_object_store: Arc<dyn ObjectStoreGetExt>,
+        sha3_digests: Arc<Mutex<DigestByBucketAndPartition>>,
+        num_part_files: usize,
+        concurrency: usize,
+    ) -> std::result::Result<(), anyhow::Error> {
+        let move_object_progress_bar = Arc::new(self.snapshot_reader.get_multi_progress().add(
+            ProgressBar::new(num_part_files as u64).with_style(
+                ProgressStyle::with_template(
+                    "[{elapsed_precise}] {wide_bar} {pos} out of {len} move object files restored ({msg})",
+                )
+                .unwrap(),
+            ),
+        ));
+
+        Abortable::new(
+            async move {
+                let sema_limit = Arc::new(Semaphore::new(concurrency));
+                let mut restore_tasks = vec![];
+
+                for (bucket, (part_num, file_metadata)) in input_files.into_iter() {
+                    let sema_limit_clone = sema_limit.clone();
+                    let epoch_dir_clone = epoch_dir.clone();
+                    let remote_object_store_clone = remote_object_store.clone();
+                    let sha3_digests_clone = sha3_digests.clone();
+                    let bar_clone = move_object_progress_bar.clone();
+                    let db = self.db.clone();
+                    let next_cp = self.next_checkpoint_after_epoch;
+
+                    let restore_task = task::spawn(async move {
+                        let mut conn = db.connect().await?;
+                        let _permit = sema_limit_clone.acquire().await.unwrap();
```

Comment:
If `try_for_each_spawned` is not a good fit, then we need to change the order of calls here -- it looks like we're going to wait for a connection in the DB pool, *then* we wait for the permit from the semaphore? That seems wrong -- it means we're going to keep hold of a DB connection and not use it until we get that permit.